api/wasm/cpp/proxy_wasm_externs.h

```diff
@@ -72,6 +72,35 @@ extern "C" uint32_t proxy_getMetadataStruct(MetadataType type, const char* name_
 extern "C" uint32_t proxy_setMetadataStruct(MetadataType type, const char* name_ptr, size_t name_size,
                                             const char* value_ptr, size_t value_size);
 
+// Generic selectors
+extern "C" uint32_t proxy_getSelectorExpression0(const char* arg0_ptr, size_t arg0_size,
```

Comment:
We can use C-style arrays to store pointers to pointers to strings. This is just slower, but is explicitly punted by WASM (see https://webassembly.org/docs/rationale/).
Leaving one function for all is also slower, since you have to push things on stack that you don't use.

Reply:
I think we should start with a simple interface which take a string "foo.bar.baz".  Paired with a static buffer to build the string in this is going to be fast and very flexible.  If it is expensive we can benchmark this ABI.  I am concerned that the V8 upcall overhead per-argument will overwhelm the memcpy cost.  We can also add some small number known metadata (enum) for common items if performance is an issue which would be faster than passing a pile of strings up from the VM.

Reply:
How do we handle keys with dots in them? For example, `request.headers` or `metadata` filter keys?

Reply:
Is `request.headers` a single selector or two? Regardless, if concatenating and splitting strings is an issue, then please use an array of strings. I don't see why it would be measurably slower than passing them separately, to be honest.

Reply:
How about using \0 as the separator?

Reply:
Fine with me (at least for the time being).

But we need to introduce complex types in the ABI for other functions anyway (e.g. we already return a list of key-values in `getHeaderMapPairs()`), and it's unclear to my why `request\0headers` (which needs to be constructed from separate strings) would be preferred over an array of strings.

Reply:
The style is all over the place. There is a custom marshalling code for maps, as you mentioned. I think we should either use a serialized form (e.g. \0 delimited) or arrays throughout. So which one do we choose?

Reply:
(Please keep in mind that we eventually want to move away from C++ for WASM. The choice must be friendly to something like Go)

Reply:
Serialized form doesn't work for all cases (e.g. large values) and it doesn't work for binary data (because `\0` can be part of the value), so let's do arrays. Is that Go-friendly?

Reply:
How about
uint8_t path_len
uint8_t segment1_len
char segment1[segment1_len]
uint8_t segment2_len
char segment2[segment2_len]
....

This can be converted into vector<string_view> very efficiently on the VM side without parsing.

Alternatively (using a wasm64 interface):

uint64_t path_len
uint64_t segment_addr1
uint64_t segment_len1
uint64_t segment_addr1
uint64_t segment_len1

However do note that for segments shorter than 15 bytes, this involves copying more bytes than if we just copied the strings themselves.

Reply:
(the second example should be segment_addr2 etc of course)

Reply:
By abuse of pointers, we convert the second into 
  absl::string_view* segment_array = (absl::string_view*)(((char*)data) + sizeof(uint64_t))

Reply:
Humm... on second thought the latter is going to be expensive since we need to validate all the pointers which means doing a range check via a virtual function WasmVm::getMemory.  That makes the array form more expensive than just copying the strings IMO